server/api/src/rsGenQuery.cpp

```diff
@@ -633,12 +649,21 @@ rsGenQuery( rsComm_t *rsComm, genQueryInp_t *genQueryInp,
         return status;
     }
 
-    // =-=-=-=-=-=-=-
-    // handle connections with community iRODS
-    if ( !zone_hint_str.empty() ) {
-        irods::error ret = proc_query_terms_for_community_server( zone_hint_str, genQueryInp );
-        if ( !ret.ok() ) {
-            irods::log( PASS( ret ) );
+    if (!zone_hint_str.empty()) {
+        // If the ZONE_KW is set, that means the client targeted a specific zone.
+        // The zone name check only applies to this case. Checking the validity of
+        // the zone name must be skipped when getZoneHintForGenQuery() derives it.
+        if (getValByKey(&genQueryInp->condInput, ZONE_KW)) {
+            if (!is_zone_name_valid(zone_hint_str)) {
+                rodsLog(LOG_ERROR, "%s:%d :: Unknown zone name [%s].", __func__, __LINE__, zone_hint_str.data());
+                return SYS_INVALID_ZONE_NAME;
+            }
+        }
```

Comment:
Thoughts?

Reply:
right, a short-circuit before the rest of the existing logic.     seems good to me.